hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OperationRunnerImpl.java

```diff
@@ -201,19 +206,28 @@ public void run(Operation op) {
     private void call(Operation op) throws Exception {
         CallStatus callStatus = op.call();
 
-        switch (callStatus) {
-            case DONE_RESPONSE:
+        switch (callStatus.ordinal()) {
+            case DONE_RESPONSE_ORDINAL:
                 handleResponse(op);
                 afterRun(op);
                 break;
-            case DONE_VOID:
-                // todo: currently there is no difference between DONE_VOID and OFFLOADED
+            case DONE_VOID_ORDINAL:
                 op.afterRun();
                 break;
-            case OFFLOADED:
+            case OFFLOADED_ORDINAL:
                 op.afterRun();
+
+                Offloaded offloaded = (Offloaded) callStatus;
+                offloaded.setNodeEngine(nodeEngine);
+                // we register the source operation before the Offloaded.start is called. So it won't happen that the
+                operationService.onStartOffloadedOperation(offloaded.source());
```

Comment:
I think it would be better to move this logic inside `Offloaded`. That way `onStartOffloadedOperation ` and `onCompletionOffloadedOperation ` will be in the same closer to each other. `offloaded.start()` can be a final method similar to `thread.start()` with all this logic inside, and another abstract method (such as `offloaded.run()`) to be implemented by subclasses.

Reply:
I made an 'init' method that takes care of this. Please check it out. It also replaces the original response handler by a decorator one that takes care of deregistration. This way no offloaded.complete is required any longer.